### PR TITLE
Parse TForce shipment documents

### PR DESCRIPTION
**Pull request by pelargir**
This updates the TForce service method to parse Bills of Lading into `ShipmentDocument` objects with binary data attached. This can be done now that the TForce API is actually returning base64-encoded PDFs.

### CODE REVIEW

File: lib/friendly_shipping/services/tforce_freight/generate_create_bol_request_hash.rb [hunk 1]
```diff
@@ -68,15 +68,15 @@ def payment(location, options)
                 name: location.company_name.presence || location.name,
                 contact: location.name,
                 email: location.email,
-                phone: { number: location.phone, },
+                phone: { number: location.phone }.compact,
```
Comment (davecandlescience) [inline lib/friendly_shipping/services/tforce_freight/generate_create_bol_request_hash.rb:71]: Will TForce accept an empty hash as the phone value?

File: lib/friendly_shipping/services/tforce_freight/generate_create_bol_request_hash.rb [hunk 2]
```diff
@@ -68,15 +68,15 @@ def payment(location, options)
...
                 address: {
                   addressLine: address_line(location),
                   city: location.city,
                   stateProvinceCode: location.region&.code,
                   postalCode: location.zip,
                   country: location.country&.code
-                }
-              },
+                }.compact
+              }.compact,
```
Comment (davecandlescience) [inline lib/friendly_shipping/services/tforce_freight/generate_create_bol_request_hash.rb:79]: Would `.compact_blank` be better here?

File: lib/friendly_shipping/services/tforce_freight/generate_create_bol_request_hash.rb (hunk 1 above)
Comment (davecandlescience) [reply lib/friendly_shipping/services/tforce_freight/generate_create_bol_request_hash.rb:71]: Wow, it sounds like they are following the bad habits of UPS.

File: lib/friendly_shipping/services/tforce_freight/generate_create_bol_request_hash.rb (hunk 2 above)
Comment (pelargir) [reply lib/friendly_shipping/services/tforce_freight/generate_create_bol_request_hash.rb:79]: Sure.

File: lib/friendly_shipping/services/tforce_freight/generate_create_bol_request_hash.rb (hunk 1 above)
Comment (pelargir) [reply lib/friendly_shipping/services/tforce_freight/generate_create_bol_request_hash.rb:71]: Apparently!

File: lib/friendly_shipping/services/tforce_freight/parse_create_bol_response.rb [hunk 3]
```diff
@@ -16,26 +16,73 @@ class << self
...
             )
           end
+
+          private
+
+          def build_cost_breakdown(rate_detail)
+            {
+              "Rates" => rate_detail["rate"].each_with_object({}) do |rate, hash|
+                hash[rate["code"]] = rate["value"]
+              end,
```
Comment (davecandlescience) [inline lib/friendly_shipping/services/tforce_freight/parse_create_bol_response.rb:81]: This is a nit but the spec confused me at first. Rates is a hash but the plural form makes it sound like it should be an array. TForce does the opposite and returns an array called "rate".

Comment (davecandlescience) [reply lib/friendly_shipping/services/tforce_freight/parse_create_bol_response.rb:81]: Their response is all mixed up re expectations from the name of the key...

Array: detail, alerts, rate, commodities
Hash: summary, detailStatus, service, shipmentCharges, shipmentWeights, timeInTransit

**Review by davecandlescience — APPROVED** [review]
Nice work! I am going to bet that you will need to Array.wrap() some of the values in the response parser. It is really hard to tell though since their key names are so mixed.

File: lib/friendly_shipping/services/tforce_freight/parse_create_bol_response.rb [hunk 4]
```diff
@@ -16,26 +16,73 @@ class << self
...
+              days_in_transit = rate_detail.dig("timeInTransit", "timeInTransit")&.to_i
+              cost_breakdown = build_cost_breakdown(rate_detail)
+            end
+
+            origin_service_center = json.dig("detail", "originServiceCenter")
+            email_sent = json.dig("detail", "pickup", "emailSent") == "true"
+            origin_is_rural = json.dig("detail", "pickup", "originIsRural") == "true"
+            destination_is_rural = json.dig("detail", "pickup", "destinationIsRural") == "true"
+
             documents = json.dig("detail", "documents", "image")&.map do |image_data|
```
[the remark this answers is not in the document]
Comment (pelargir) [reply lib/friendly_shipping/services/tforce_freight/parse_create_bol_response.rb:48]: I haven't noticed that happening yet.

File: lib/friendly_shipping/services/tforce_freight/parse_create_bol_response.rb (hunk 3 above)
Comment (pelargir) [reply lib/friendly_shipping/services/tforce_freight/parse_create_bol_response.rb:81]: Yep. Fun, isn't it?

**Pull request merged**